### PR TITLE
Add a 'flow' option that only checks files annotated @flow

### PR DESCRIPTION
**Pull request by nicks**
Fixes https://github.com/gajus/eslint-plugin-flowtype/issues/5

### CODE REVIEW

File: src/rules/requireReturnType.js [hunk 1]
```diff
@@ -1,6 +1,12 @@
 import _ from 'lodash';
+import {isFlowFile} from './../utilities';
```
Comment (gajus) [inline src/rules/requireReturnType.js]: Change to:

```js
import {
    isFlowFile
} from './../utilities';
```

File: src/rules/requireReturnType.js [hunk 2]
```diff
@@ -1,6 +1,12 @@
...
 
 export default (context) => {
+    const checkThisFile = !context.settings['flowtype:onlyFlowFiles'] || isFlowFile(context);
```
Comment (gajus) [inline src/rules/requireReturnType.js]: Change to (note a new empty line and a different variable name):

```js
const checkThisFile = _.get(context, 'settings.flowtype.onlyFilesWithFlowAnnotation') || isFlowFile(context);

if (!checkThisFile) {
    return () => {}
}
```

File: tests/rules/assertions/requireParameterType.js [hunk 3]
```diff
@@ -47,6 +47,17 @@ export default {
...
+        },
+        {
+            code: '// @flow\n(foo) => {}',
+            errors: [
+                {
+                    message: 'Missing "foo" parameter type annotation.'
+                }
+            ],
+            settings: {
+              'flowtype:onlyFlowFiles': true
```
Comment (gajus) [inline tests/rules/assertions/requireParameterType.js]: This becomes:

```js
settings: {
    flowtype: {
        onlyFilesWithFlowAnnotation: true
    }
}
```

File: tests/rules/assertions/requireParameterType.js [hunk 4]
```diff
@@ -64,6 +75,12 @@ export default {
         },
         {
             code: '([foo]: Array) => {}'
+        },
+        {
+            code: '(foo) => {}',
+            settings: {
+              'flowtype:onlyFlowFiles': true
```
Comment (gajus) [inline tests/rules/assertions/requireParameterType.js]: This becomes:

```js
settings: {
    flowtype: {
        onlyFilesWithFlowAnnotation: true
    }
}
```

File: tests/rules/assertions/requireReturnType.js [hunk 5]
```diff
@@ -134,6 +134,34 @@ export default {
...
+        },
+        {
+            code: '// @flow\n(foo) => { return 1; }',
+            errors: [
+                {
+                    message: 'Missing return type annotation.'
+                }
+            ],
+            settings: {
+              'flowtype:onlyFlowFiles': true
```
Comment (gajus) [inline tests/rules/assertions/requireReturnType.js]: This becomes:

```js
settings: {
    flowtype: {
        onlyFilesWithFlowAnnotation: true
    }
}
```

File: tests/rules/assertions/requireReturnType.js [hunk 6]
```diff
@@ -134,6 +134,34 @@ export default {
...
+        },
+        {
+            code: '// @flow\n (foo) => { return undefined; }',
+            errors: [
+                {
+                    message: 'Must annotate undefined return type.'
+                }
+            ],
+            settings: {
+              'flowtype:onlyFlowFiles': true
```
Comment (gajus) [inline tests/rules/assertions/requireReturnType.js]: This becomes:

```js
settings: {
    flowtype: {
        onlyFilesWithFlowAnnotation: true
    }
}
```

File: tests/rules/assertions/requireReturnType.js [hunk 7]
```diff
@@ -217,6 +245,27 @@ export default {
...
+            },
+            options: [
+                'always'
+            ]
+        },
+        {
+            code: '(foo) => { return undefined; }',
+            settings: {
+              'flowtype:onlyFlowFiles': true
```
Comment (gajus) [inline tests/rules/assertions/requireReturnType.js]: This becomes:

```js
settings: {
    flowtype: {
        onlyFilesWithFlowAnnotation: true
    }
}
```

File: tests/rules/assertions/requireReturnType.js [hunk 8]
```diff
@@ -217,6 +245,27 @@ export default {
                     annotateUndefined: 'always'
                 }
             ]
+        },
+        {
+            code: '(foo) => { return 1; }',
+            settings: {
+              'flowtype:onlyFlowFiles': true
```
Comment (gajus) [inline tests/rules/assertions/requireReturnType.js]: This becomes:

```js
settings: {
    flowtype: {
        onlyFilesWithFlowAnnotation: true
    }
}
```

**Comment by gajus** [comment]
Finally, please include a mention of `flowtype.onlyFilesWithFlowAnnotation` setting in the documentation.

**Comment by nicks** [comment]
Done! All comments addressed. and squashed.

**Comment by gajus** [comment]
Thank you for the PR and for sticking through the amendments. Will publish an update in few minutes.

**Comment by gajus** [comment]
In case you'd contribute in future, please note that the documentation file (`./README.md`) is built from files in `./.README/` using `npm run readme`.

This is not documented, therefore I made the change myself (https://github.com/gajus/eslint-plugin-flowtype/commit/9557226d1cb278ad454cc33df6fcd9245801dd51) after merging your changes.

**Comment by gajus** [comment]
v2.2.6 has been published.

However, I don't like the implementation of [`./src/utilities/isFlowFile.js`](https://github.com/gajus/eslint-plugin-flowtype/blob/a545129fd54f19ddad7d07a57a816b6ff601c1b8/src/utilities/isFlowFile.js). I have asked in the flowtype repository (https://github.com/facebook/flow/issues/1641) to clarify what should be the logic for checking existence of `@flow` comment.

**Comment by nicks** [comment]
sounds good, thanks!

**Comment by yavuzmester** [comment]
Aside from putting @flow to the beginning of the file, it would be good if I can tag some of my functions in that file. For example I may have a file named utils.js and inside I may have a bunch of functions. Then I may tag some of those functions to be checked by flow. Then if I have another file with @flow tag at the beginning, and invoking the tagged functions in utils.js from that file may be checked by Flow. How does it sound?

Because I do not want to create a new file and move the functions that I want to check to that file.

**Comment by yavuzmester** [comment]
Aside from putting @flow to the beginning of the file, it would be good if I can tag some of my functions in that file. For example I may have a file named utils.js and inside I may have a bunch of functions. The utils.js file may not have a @flow tag at the beginning, but I may tag some of the functions to be checked by flow. Then if I have another file with @flow tag at the beginning, and invoking the tagged functions in utils.js from that file may be checked by Flow.

Because I do not want to create a new file and move the functions that I want to check to that file. How does it sound?

**Comment by gajus** [comment]
We are going to stick with flowtype implementation. I suggest to raise your idea with flowtype codebase.

**Comment by yavuzmester** [comment]
Cool, thank you.

**Pull request merged**